README.md

# ImagePicker

[![Download](https://api.bintray.com/packages/wrs/maven/ImagePicker/images/download.svg)](https://bintray.com/wrs/maven/ImagePicker/_latestVersion)

## 背景

开发App过程中经常会有选择图片、裁剪图片的需求。实现该需求最快的方式就是通过隐式调用来调起系统提供的相机、图库UI、文档UI来进行选择，通过`ACTION`为`com.android.camera.action.CROP`的隐式调用来调起裁剪。

## 简介

该库提供快速从系统相机、图库UI或文档UI选择图片并裁剪的功能，包含Android 6.0开始动态权限申请（文件存储权限）以及7.0文件协议（FileProvider）的处理。

核心功能拍照、选图、裁剪可自由搭配使用或单独使用，同时未内嵌UI，项目可高度定制自己的UI。

Demo介绍了常用的使用方式：

![简介](screenshot/简介.png)

## 主要功能

1.选图功能：从相机、图库UI、文档UI选择图片；

2.裁剪功能；

3.选图+裁剪组合使用；

## 使用方式

`build.gradle`添加依赖：

```groovy
// 确保仓库已添加jcenter()
allprojects {
    repositories {
        ...
        jcenter()
    }
}

// 添加依赖
dependencies {
	compile 'com.lancewu:ImagePicker:<version>' 
}
```

`<version>`请替换为对应的版本号。

### e.g.拍照+裁剪

需求：从相机拍照并裁剪得到一张200*400的图片；

实现代码如下：

```java
public void clickCameraCrop(View view) {
    // 从相机拍照并裁剪
    File file = new File(getExternalCacheDir(), "camera_crop.jpg");
    // 创建裁剪参数
    ImagePicker.CropConfigBuilder cropConfigBuilder = new ImagePicker.CropConfigBuilder()
            .aspect(1, 2) // 比例1：2
            .outputSize(200, 400) // 输出大小200*400
            .outputFile(file);  // 最终文件保存路径
    // 创建选择器
    mPicker = new ImagePicker.Builder(this)
            .fromCamera(file) // 表示从相机选择，并设置拍照保存文件
            .withCrop(cropConfigBuilder) // 拍照完紧接裁剪
            .build();
    // 调用选图
    mPicker.pick(mCallback);
}

private OnImagePickerCallback mCallback = new OnImagePickerCallback() {
    @Override
    public void onPickError(@ErrorCode int errorCode) {
        // 发生错误，具体错误参考：@ErrorCode
        showToast("ImagePicker-发生错误：" + errorCode);
    }

    @Override
    public void onPickSuccess(@NonNull ImagePickerResult result) {
        // 选图/裁剪回调
        InputStream inputStream = null;
        try {
            // 从选择结果中取出文件Uri，进行想要的处理，这边直接显示
            inputStream = getContentResolver().openInputStream(result.getImageUri());
            Bitmap bitmap = BitmapFactory.decodeStream(inputStream);
            mImageView.setImageBitmap(bitmap);
            String text = "图片信息：宽*高=" + bitmap.getWidth() + "*" +bitmap.getHeight();
            mInfoTv.setText(text);
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } finally {
            StreamUtils.close(inputStream);
        }
    }

    @Override
    public void onPickCancel() {
        // 主动取消选择/裁剪时回调
        showToast("ImagePicker-取消选择");
    }

    void showToast(String msg) {
        Toast.makeText(MainActivity.this, msg, Toast.LENGTH_SHORT).show();
    }
};
```

更详细的使用方式可参照Demo。

### 接口API

[API概览](https://wurensen.github.io/ImagePicker/docs/index.html) 

[核心API-选择功能设置：ImagePicker.Builder](https://wurensen.github.io/ImagePicker/docs/com/lancewu/imagepicker/ImagePicker.Builder.html) 

[核心API-裁剪参数设置：ImagePicker.CropConfigBuilder](https://wurensen.github.io/ImagePicker/docs/com/lancewu/imagepicker/ImagePicker.CropConfigBuilder.html) 

## Change Log

[Change Log](CHANGELOG.md)

## License

```txt
Copyright 2019 LanceWu

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
# Change Log

## Version 0.0.3(26/2/2019)

1.Bugfix：修复多个app同时接入该库时内置的`ImagePickerFileProvider`属性`authorities`重复问题；

## Version 0.0.2(18/2/2019)

1.裁剪增加noFaceDetection参数设置（默认禁用，是否有效取决于裁剪页面自身是否支持）；

## Version 0.0.1(13/2/2019)

1.选图功能：从相机、图库UI、文档UI选择图片；

2.裁剪功能；

3.选图+裁剪组合使用；

app/src/main/java/com/lancewu/imagepickerdemo/MainActivity.java
package com.lancewu.imagepickerdemo;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.os.Environment;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.lancewu.imagepicker.ImagePicker;
import com.lancewu.imagepicker.ImagePickerResult;
import com.lancewu.imagepicker.OnImagePickerCallback;
import com.lancewu.imagepicker.util.StreamUtils;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;

public class MainActivity extends AppCompatActivity {

    private ImageView mImageView;
    private TextView mInfoTv;
    private ImagePicker mPicker;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        mImageView = findViewById(R.id.iv);
        mInfoTv = findViewById(R.id.info_tv);
    }

    public void clickCameraCrop(View view) {
        // 从相机拍照并裁剪
        File file = new File(getExternalCacheDir(), "camera_crop.jpg");
        // 创建裁剪参数
        ImagePicker.CropConfigBuilder cropConfigBuilder = new ImagePicker.CropConfigBuilder()
                .aspect(1, 2) // 比例1：2
                .outputSize(200, 400) // 输出大小200*400
                .outputFile(file);  // 最终文件保存路径
        // 创建选择器
        mPicker = new ImagePicker.Builder(this)
                .fromCamera(file) // 表示从相机选择，并设置拍照保存文件
                .withCrop(cropConfigBuilder) // 拍照完紧接裁剪
                .build();
        // 调用选图
        mPicker.pick(mCallback);
    }

    public void clickGalleryCrop(View view) {
        File file = new File(getExternalCacheDir(), "gallery_crop.jpg");
        ImagePicker.CropConfigBuilder cropConfigBuilder = new ImagePicker.CropConfigBuilder()
                .aspect(1, 1)
//                .outputSize(100, 100) // 采用默认大小200
                .outputFile(file);
        mPicker = new ImagePicker.Builder(this)
                .fromGallery()
                .withCrop(cropConfigBuilder)
                .build();
        mPicker.pick(mCallback);
    }

    public void clickDocumentCrop(View view) {
        File file = new File(getExternalCacheDir(), "document_crop.jpg");
        ImagePicker.CropConfigBuilder cropConfigBuilder = new ImagePicker.CropConfigBuilder()
                .aspect(1, 1) // 比例
                .outputSize(300, 300) // 裁剪输出尺寸
                .faceDetection(true) // 需要人脸识别
                .outputFile(file); // 裁剪存储的文件
        mPicker = new ImagePicker.Builder(this)
                .fromDocument()
                .withCrop(cropConfigBuilder)
                .build();
        mPicker.pick(mCallback);
    }

    public void clickCamera(View view) {
//        File file = new File(getExternalCacheDir(), "camera.jpg");
        // SD卡需要动态申请权限
        File file = new File(Environment.getExternalStorageDirectory(), "camera.jpg");
        mPicker = new ImagePicker.Builder(this)
                .fromCamera(file)
                .build();
        mPicker.pick(mCallback);
    }

    public void clickGallery(View view) {
        mPicker = new ImagePicker.Builder(this)
                .fromGallery()
                .build();
        mPicker.pick(mCallback);
    }

    public void clickDocument(View view) {
        mPicker = new ImagePicker.Builder(this)
                .fromDocument()
                .build();
        mPicker.pick(mCallback);
    }

    public void clickCrop(View view) {
        InputStream inputStream = null;
        FileOutputStream fileOutputStream = null;
        File file = null;
        try {
            inputStream = getAssets().open("justCrop.jpg");
            file = new File(getExternalCacheDir(), "crop_input.jpg");
            fileOutputStream = new FileOutputStream(file);
            final byte[] buf = new byte[8 * 1024];
            int read;
            while ((read = inputStream.read(buf)) != -1) {
                fileOutputStream.write(buf, 0, read);
            }
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            try {
                if (inputStream != null) {
                    inputStream.close();
                }
                if (fileOutputStream != null) {
                    fileOutputStream.close();
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        if (file == null) {
            return;
        }
        File cropFile = new File(getExternalCacheDir(), "only_crop_output.jpg");
        ImagePicker.CropConfigBuilder cropConfigBuilder = new ImagePicker.CropConfigBuilder()
                .aspect(1, 1)
                .outputSize(700, 700)
                .inputFile(file)
                .outputFile(cropFile);
        mPicker = new ImagePicker.Builder(this)
                .withCrop(cropConfigBuilder)
                .build();
        mPicker.pick(mCallback);
    }

    private OnImagePickerCallback mCallback = new OnImagePickerCallback() {
        @Override
        public void onPickError(@ErrorCode int errorCode) {
            // 发生错误，具体错误参考：@ErrorCode
            showToast("ImagePicker-发生错误：" + errorCode);
        }

        @Override
        public void onPickSuccess(@NonNull ImagePickerResult result) {
            // 选图/裁剪回调
            InputStream inputStream = null;
            try {
                // 从选择结果中取出文件Uri，进行想要的处理，这边直接显示
                inputStream = getContentResolver().openInputStream(result.getImageUri());
                Bitmap bitmap = BitmapFactory.decodeStream(inputStream);
                mImageView.setImageBitmap(bitmap);
                String text = "图片信息：宽*高=" + bitmap.getWidth() + "*" + bitmap.getHeight();
                mInfoTv.setText(text);
            } catch (FileNotFoundException e) {
                e.printStackTrace();
            } finally {
                StreamUtils.close(inputStream);
            }
        }

        @Override
        public void onPickCancel() {
            // 主动取消选择/裁剪时回调
            showToast("ImagePicker-取消选择");
        }

        void showToast(String msg) {
            Toast.makeText(MainActivity.this, msg, Toast.LENGTH_SHORT).show();
        }
    };
}
